### PR TITLE
Ensure section renumbering tool handles removing clauses.

### PR DESCRIPTION
**Pull request by BillWagner**
In preparation for **Recommendation 5** in #425

This PR updates the section renumbering tool to handle situations where we remove a clause. (I think it will handle adding a clause as well, but that wasn't tested.)

The focus of the change is to remove the hardcoded list of files from the `Program.cs` file. Instead, read the list of frontmatter, main body clauses, and annexes from a .json config file. The config file is stored in the same folder as the standard text.

- If you remove a file from the standard, and don't remove it from the config file, the dry run fails. That should help us catch any instances when the PR is opened.
- The dry run will catch any links in the text that resolved to a clause in the removed section. 
- The full run renumbers headers and all links.

### CODE REVIEW

File: tools/StandardAnchorTags/TocSectionNumberBuilder.cs [hunk 1]
```diff
@@ -33,13 +33,13 @@ private struct SectionHeader
...
-        private StringBuilder tocContent = new StringBuilder();
-        private readonly Dictionary<string, SectionLink> sectionLinkMap = new Dictionary<string, SectionLink>();
+        private readonly StringBuilder tocContent = new();
+        private readonly Dictionary<string, SectionLink> sectionLinkMap = new();
         private bool isAnnexes;
 
         // Running array of entries for the current headings. 
         // Starting with H1 is headings[0], H2 is headings[1] etc.
-        int[] headings = new int[8];
+        readonly int[] headings = new int[8];
```
Comment (jskeet) [inline tools/StandardAnchorTags/TocSectionNumberBuilder.cs]: Add an explicit `private` as well for consistent?

File: tools/StandardAnchorTags/TocSectionNumberBuilder.cs [hunk 2]
```diff
@@ -53,12 +53,24 @@ public TocSectionNumberBuilder(string pathFromToolToStandard, bool dryRun)
...
+        /// <param name="fileName">The front matter file name.</param>
+        /// <remarks>
+        /// The front matter entries don't add section numbers in the headers.
+        /// Otherwise, this is the same logic for the main file.
+        /// </remarks>
+        public async Task AddFrontMatterTocEntries(string fileName)
         {
-            foreach (var entry in entries)
+            string? line;
+            using var stream = new StreamReader($"{PathToStandardFiles}/{fileName}");
```
Comment (jskeet) [inline tools/StandardAnchorTags/TocSectionNumberBuilder.cs]: Maybe use Path.Combine here? (I'm sure it's fine either way... just my habit.)
(I also generally use `File.OpenText` but it's a personal preference thing...)

File: tools/StandardAnchorTags/TocSectionNumberBuilder.cs [hunk 3]
```diff
@@ -53,12 +53,24 @@ public TocSectionNumberBuilder(string pathFromToolToStandard, bool dryRun)
...
+            while ((line = await stream.ReadLineAsync()) != null)
             {
-                tocContent.AppendLine($"- [{entry.linkText}]({entry.fileName})");
+                if (line.StartsWith("# "))
+                {
+                    var linkText = line[2..];
+                    tocContent.AppendLine($"- [{linkText}]({fileName})");
+                    // Done: return.
+                    return;
+                }
```
Comment (jskeet) [inline tools/StandardAnchorTags/TocSectionNumberBuilder.cs:72]: Do we want to throw an exception if we get to the end of this method? It means we haven't found a heading, which is presumably a problem.

Also, can we not just read the first line of the file and assert that it starts with `# `? (Again, if it doesn't that's presumably something wrong...)

File: tools/Utilities/Clauses.cs [hunk 4]
```diff
@@ -0,0 +1,9 @@
+﻿namespace Utilities
+{
+    public class Clauses
+    {
+        public string[] FrontMatter { get; set; } = Array.Empty<string>();
+        public string[] MainBody { get; set; } = Array.Empty <string>();
```
Comment (jskeet) [inline tools/Utilities/Clauses.cs]: Nit: remove space after `Empty`.
(I'm interested in your preference for arrays over `List<string>` here, but that can be something we discuss at leisure...)

**Review by jskeet — APPROVED** [review]
Fine to merge as-is, but with a few optional suggestions.

File: tools/StandardAnchorTags/TocSectionNumberBuilder.cs (hunk 2 above)
Comment (BillWagner) [reply tools/StandardAnchorTags/TocSectionNumberBuilder.cs]: Good suggestion. I applied both.

File: tools/StandardAnchorTags/TocSectionNumberBuilder.cs (hunk 3 above)
Comment (BillWagner) [reply tools/StandardAnchorTags/TocSectionNumberBuilder.cs:72]: Both good suggestions. (I thought about the 2nd, but then wondered if we might ever add comments above the H1. I tend to agree that we won't do that).

File: tools/Utilities/Clauses.cs (hunk 4 above)
Comment (BillWagner) [reply tools/Utilities/Clauses.cs]: I suppose I could say something brilliant about these never being resized, only set during serialization, etc.

But, the truth is I used Visual Studio's "paste JSON text as class" and used that definition. 🔨

**Pull request merged**